transaction/core/src/ring_signature/rct_bulletproofs.rs

```diff
@@ -347,46 +560,70 @@ fn sign_with_balance_check<CSPRNG: RngCore + CryptoRng>(
         }
     }
 
-    let pseudo_output_commitments: Vec<CompressedCommitment> = commitments
-        .iter()
-        .take(num_inputs)
-        .map(CompressedCommitment::from)
+    // The actual pseudo output commitments use the blindings from
+    // `pseudo_output_blinding` and not the original true input.
+    let pseudo_output_commitments: Vec<CompressedCommitment> = pseudo_output_commitments
+        .into_iter()
+        .map(|point| CompressedCommitment::from(&point.compress()))
         .collect();
 
     // Extend the message with the range proof and pseudo_output_commitments.
     // This ensures that they are signed by each RingMLSAG.
-    let range_proof_bytes = range_proof.to_bytes();
     let extended_message_digest = compute_extended_message_either_version(
         block_version,
         message,
         &pseudo_output_commitments,
-        &range_proof_bytes,
+        &range_proof,
+        &range_proofs,
     );
 
     // Prove that the signer is allowed to spend a public key in each ring, and that
     // the input's value equals the value of the pseudo_output.
     let mut ring_signatures: Vec<RingMLSAG> = Vec::new();
     for i in 0..num_inputs {
         let real_index = real_input_indices[i];
-        let (onetime_private_key, value, blinding) = input_secrets[i];
+        let input_secret = &input_secrets[i];
+        let generator = generator_cache.get(input_secret.token_id);
         let ring_signature = RingMLSAG::sign(
             &extended_message_digest,
             &rings[i],
             real_index,
-            &onetime_private_key,
-            value,
-            &blinding,
+            &input_secret.onetime_private_key,
+            input_secret.value,
+            &input_secret.blinding,
             &pseudo_output_blindings[i],
-            &generator,
+            generator,
             rng,
         )?;
         ring_signatures.push(ring_signature);
     }
 
+    let mut pseudo_output_token_ids: Vec<u64> = input_secrets
+        .iter()
+        .map(|secret| *secret.token_id)
+        .collect();
+    let mut output_token_ids: Vec<u64> = output_secrets
+        .iter()
+        .map(|secret| *secret.token_id)
+        .collect();
+
+    if !block_version.mixed_transactions_are_supported() {
+        pseudo_output_token_ids.clear();
+        output_token_ids.clear();
+        assert!(!range_proof.is_empty());
+        assert!(range_proofs.is_empty());
+    } else {
+        assert!(range_proof.is_empty());
+        assert!(!range_proofs.is_empty());
```

Comment:
My take is that this code runs on clients, I think it is acceptable that it panics instead of produces an unexpected result.

Reply:
I am okay with either way

In the future I hope that we will drop support for older block versions that are no longer relevant, and then we can delete the branches having to do with `range_proof`, so ultimately there is a path to get rid of this.

Reply:
@remoun maybe we can leave it as assert for now, so that when SDKs are testing with it, it's a hard assert, but we can make a ticket so that before we branch for release we change it to debug assert?

Reply:
SGTM